src/EditorFeatures/Core/Implementation/CallHierarchy/Finders/AbstractCallFinder.cs

```diff
@@ -48,51 +50,64 @@ public void CancelSearch()
         {
-            Task.Run(() => SearchAsync(callback, searchScope, _cancellationSource.Token), _cancellationSource.Token);
-        }
-
-        private async Task SearchAsync(ICallHierarchySearchCallback callback, CallHierarchySearchScope scope, CancellationToken cancellationToken)
-        {
-            callback.ReportProgress(0, 1);
-
             var asyncToken = _asyncListener.BeginAsyncOperation(this.GetType().Name + ".Search");
 
-            // Follow the search task with task that lets the callback know we're done and which
-            // marks the async operation as complete.  Note that we pass CancellationToken.None
-            // here.  That's intentional.  This operation is *not* cancellable.
+            // NOTE: This task has CancellationToken.None specified, since it must complete no matter what
+            // so the callback is appropriately notified that the search has terminated.
+            Task.Run(async () =>
+            {
+                // The error message to show if we had an error. null will mean we succeeded.
+                string completionErrorMessage = null;
+                try
+                {
+                    await SearchAsync(workspace, searchScope, callback, _cancellationSource.Token).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    completionErrorMessage = EditorFeaturesResources.Canceled;
+                }
+                catch (Exception e) when (FatalError.ReportWithoutCrash(e))
+                {
+                    completionErrorMessage = e.Message;
+                }
+                finally
+                {
+                    if (completionErrorMessage != null)
+                    {
+                        callback.SearchFailed(completionErrorMessage);
+                    }
+                    else
+                    {
+                        callback.SearchSucceeded();
+                    }
+                }
+            }, CancellationToken.None).CompletesAsyncOperation(asyncToken);
+        }
 
-            var workspace = _project.Solution.Workspace;
-            var currentProject = workspace.CurrentSolution.GetProject(_project.Id);
-            var compilation = await currentProject.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
-            var resolution = _symbol.Resolve(compilation, cancellationToken: cancellationToken);
+        private async Task SearchAsync(Workspace workspace, CallHierarchySearchScope scope, ICallHierarchySearchCallback callback, CancellationToken cancellationToken)
+        {
+            var project = workspace.CurrentSolution.GetProject(_projectId);
+            
+            if (project == null)
+            {
+                throw new Exception(string.Format(WorkspacesResources.The_symbol_0_cannot_be_located_within_the_current_solution, SymbolName));
```

Comment:
If we'r egoing to throw our own exceptions, i would *really* prefer it was our own exception type.